crypto/dilithium/p_dilithium3.c

```diff
@@ -9,117 +9,300 @@
+    PQDSA_KEY_clear(key);
+    return 0;
+  }
   return 1;
+}
 
-err:
-  if (key != NULL) {
-    OPENSSL_free(key->pub);
-    OPENSSL_free(key->priv);
+void PQDSA_KEY_free(PQDSA_KEY *key) {
+  if (key == NULL) {
+    return;
   }
+  PQDSA_KEY_clear(key);
   OPENSSL_free(key);
-  return 0;
+}
 
+const PQDSA *PQDSA_KEY_get0_sig(PQDSA_KEY* key) {
+  return key->pqdsa;
 }
 
-static int pkey_dilithium3_sign_message(EVP_PKEY_CTX *ctx, uint8_t *sig,
-                                        size_t *siglen, const uint8_t *tbs,
-                                        size_t tbslen) {
-  if (ctx == NULL || ctx->pkey->pkey.ptr == NULL) {
-    OPENSSL_PUT_ERROR(EVP, EVP_R_MISSING_PARAMETERS);
+// PQDSA PKEY functions
+
+static int pkey_pqdsa_init(EVP_PKEY_CTX *ctx) {
+  PQDSA_PKEY_CTX *dctx;
+  dctx = OPENSSL_zalloc(sizeof(PQDSA_PKEY_CTX));
+  if (dctx == NULL) {
     return 0;
   }
 
-  DILITHIUM3_KEY *key = ctx->pkey->pkey.ptr;
+  ctx->data = dctx;
 
-  if (!key->priv) {
-    OPENSSL_PUT_ERROR(EVP, EVP_R_NOT_A_PRIVATE_KEY);
+  return 1;
+}
+
+static void pkey_pqdsa_cleanup(EVP_PKEY_CTX *ctx) {
+  OPENSSL_free(ctx->data);
+}
+
+static int pkey_pqdsa_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
+  GUARD_PTR(ctx);
+  PQDSA_PKEY_CTX *dctx = ctx->data;
+  GUARD_PTR(dctx);
+  const PQDSA *pqdsa = dctx->pqdsa;
+  if (pqdsa == NULL) {
+    if (ctx->pkey == NULL) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+      return 0;
+    }
+    pqdsa = PQDSA_KEY_get0_sig(ctx->pkey->pkey.pqdsa_key);
+  }
+
+  PQDSA_KEY *key = PQDSA_KEY_new();
+  if (key == NULL ||
+      !PQDSA_KEY_init(key, pqdsa) ||
+      !pqdsa->method->keygen(key->public_key, key->secret_key) ||
+      !EVP_PKEY_set_type(pkey, EVP_PKEY_NISTDSA)) {
+    PQDSA_KEY_free(key);
     return 0;
+      }
+
+  pkey->pkey.pqdsa_key = key;
+  return 1;
+}
+
+static int pkey_pqdsa_sign_message(EVP_PKEY_CTX *ctx, uint8_t *sig,
+                                        size_t *siglen, const uint8_t *tbs,
+                                        size_t tbslen) {
+  PQDSA_PKEY_CTX *dctx = ctx->data;
+  const PQDSA *pqdsa = dctx->pqdsa;
+  if (pqdsa == NULL) {
+    if (ctx->pkey == NULL) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+      return 0;
+    }
+    pqdsa = PQDSA_KEY_get0_sig(ctx->pkey->pkey.pqdsa_key);
   }
+
   // Caller is getting parameter values.
   if (sig == NULL) {
-    *siglen = DILITHIUM3_SIGNATURE_BYTES;
+    *siglen = pqdsa->signature_len;
     return 1;
   }
 
-  if (*siglen < DILITHIUM3_SIGNATURE_BYTES) {
+  if (*siglen < pqdsa->signature_len) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
     return 0;
   }
 
-  if (ml_dsa_65_sign(sig, siglen, tbs, tbslen, NULL, 0, key->priv) != 0) {
+  // Check that the context is properly configured.
+  if (ctx->pkey == NULL ||
+      ctx->pkey->pkey.pqdsa_key == NULL ||
+      ctx->pkey->type != EVP_PKEY_NISTDSA) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATON_NOT_INITIALIZED);
+    return 0;
+      }
```

Comment:
```suggestion
  }
```

Reply:
fixed in [e2234e3](https://github.com/aws/aws-lc/pull/1963/commits/e2234e347025d64cd9f912fb78555585e03a6198)